New plot_type 1d_profile in monitor

FranziskaWinterstein opened the pull request:
## Description

This implements a new one dimensional profile plot option '1d_profile' in the diagnostic monitor/multi_datasets.py. It allows one dimensional  preprocessed data with only dimension height or pressure and plots all datasets in one figure. It is possible to plot the absolute values or differences. The latter as absolute differences or relative differences.

To prevent misunderstanding the former plot_type 'profile' is renamed to 'zonal_mean_profile'. Backward compatibility is given until the option 'profile' can be removed in future versions.

- Closes #3158 
- Link to doc: https://esmvaltool--3178.org.readthedocs.build/en/3178/recipes/recipe_monitor.html#example-plots

Example:
![1d_profile](https://github.com/ESMValGroup/ESMValTool/assets/119339136/bee08291-644a-405b-b7f4-8c7adb5c5775)



* * *

## Before you get started

- [x] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [x] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [x] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [x] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [x] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

***

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -390,11 +439,13 @@ def __init__(self, config):
-            if plot_type == 'profile':
+            # Defaults for zonal_mean_profile plots
+            if plot_type in ['zonal_mean_profile', 'profile', '1d_profile']:
                 self.plots[plot_type].setdefault('log_y', True)
```
Don't forget to set a default for `log_x` (`False`) and `aspect_ratio` (`1.5`) for `1d_profile` here

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -292,6 +298,47 @@
+aspect_ratio: float, optional
+    Optional set the aspect ratio of the one-dimensional plot. Default is 3/2,
+    which results in a slender upright plot.
+y_log: bool, optional
+    Optional set a logarithmic scale for the vertical axis
+x_log: bool, optional
+    Optional set a logarithmic scale for the x-axis
```
```suggestion
log_x: bool, optional (default: False)
    Use logarithmic X-axis.
log_y: bool, optional (default: True)
    Use logarithmic Y-axis.
```

This is called `log_x` and `log_y` in the code

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -574,9 +626,8 @@ def _get_cbar_label(self, plot_type, dataset, bias=False):
         return cbar_label
 
-    def _get_gridline_kwargs(self):
+    def _get_gridline_kwargs(self, plot_type='map'):
```
```suggestion
    def _get_gridline_kwargs(self, plot_type):
```

I think it's better to make that non-optional, a default value `map` is not really reasonable

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -826,10 +877,12 @@ def _plot_map_without_ref(self, plot_func, dataset):
-        logger.info("Plotting profile with reference dataset '%s' for '%s'",
+    def _plot_zonal_mean_profile_with_ref(self, plot_func, dataset,
+                                          ref_dataset):
+        """Plot zonal mean plot for single dataset with a reference dataset."""
```
```suggestion
        """Plot zonal mean profile for single dataset with reference dataset."""
```

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -931,10 +984,11 @@ def _plot_profile_with_ref(self, plot_func, dataset, ref_dataset):
-        plot_type = 'profile'
-        logger.info("Plotting profile without reference dataset for '%s'",
+    def _plot_zonal_mean_profile_without_ref(self, plot_func, dataset):
+        """Plot zonal mean for single dataset without a reference dataset."""
```
```suggestion
        """Plot zonal mean profile for single dataset without reference dataset."""
```

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1284,12 +1341,18 @@ def create_map_plot(self, datasets, short_name):
+        if plot_type not in self.plots and 'profile' not in self.plots:
             return
 
+        if 'profile' in self.plots:
+            logger.warning("Plot type 'profile' is deprecated and will be"
+                           " removed in version 2.10. It was renamed to"
+                           " 'zonal_mean_profile'")
+            self.plots['zonal_mean_profile'] = self.plots['profile']
+
```
Can you move this part to the `__init__` function right below line 408 (`# Check given plot types and set default settings for them`)? This would simplify the remaining code a lot.

You can simply replace `profile` by `zonal_mean_profile` once in the beginning and then do not have to consider `profile` anywhere in the code.

E.g.,

```py
if 'profile' in self.plots:
    # show warning
    self.plots['zonal_mean_profile'] = self.plots.pop('profile')
```

FranziskaWinterstein replied:
@schlunma Yes, that is a neat solution. Do you think it is a problem, that the options of `zonal_mean_profile` get overwritten by `profile` if both plotting types are used?

schlunma replied:
Very good point! I think it's best if you raise an error when both options are used (you can check that right above this line), it doesn't make any sense to use both.

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -292,6 +298,47 @@
+
+Configuration options for plot type ``1d_profile``
+--------------------------------------------------
+plot_bias: string, optional
+    Optional switch to choose whether difference plot to reference should
+    be applied. And if plot_bias is not False or empty, it defines if
+    difference is 'absolute' or 'relative'.
```
I would recommend not doing this in the diagnostic but rather use the [`bias`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html#id36) preprocessor. That way the code for the diagnostic will be much simpler.

schlunma replied:
Yes, you should be able to use the [`convert_units`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html#convert-units) for this. Note that you also need to specify [`custom_order: true`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/overview.html#recipe-section-preprocessors) in your preprocessor since in the [default order](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#preprocessor-functions) `bias` is executed after `convert_units`.

FranziskaWinterstein replied:
Thank you, good to know. Then I will delete the bias calculation in the 1d_profile.

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1360,14 +1424,141 @@ def create_profile_plot(self, datasets, short_name):
+        if not datasets:
+            raise ValueError(f"No input data to plot '{plot_type}' given")
+
+        # Get reference dataset if possible
+        ref_dataset = self._get_reference_dataset(datasets, short_name)
+        if ref_dataset is None:
+            logger.info("Plotting %s without reference dataset", plot_type)
+            ref_cube = None
+        else:
+            logger.info("Plotting %s with reference dataset '%s'", plot_type,
+                        self._get_label(ref_dataset))
+            ref_cube = ref_dataset['cube']
```
This part should not be necessary of you use the `bias` preprocessor.

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1360,14 +1424,141 @@ def create_profile_plot(self, datasets, short_name):
+        if gridline_kwargs is not False:
+            axes.grid(**gridline_kwargs)
+        # nicer aspect ratio
+        aspect_ratio = self.plots[plot_type].get('aspect_ratio', 3 / 2)
```
```suggestion
        aspect_ratio = self.plots[plot_type]['aspect_ratio']
```

If default is set properly in `__init__`

schlunma reviewed (commented):
I just had a brief look at the code, looks very good already!! :+1: 

My two main comments:

- At the very beginning of the diagnostic (in the `__init__` method), you can add a check for `plot_type == 'profile'` and replace this if necessary with
  ```py
  self.plots['zonal_mean_profile'] = self.plots.pop('profile')
  ```
  Then you can basically remove all further appearances of `profile` in the remaining code, which makes it much simpler.
- I would recommend to not add a separate `plot_bias` option here but rather use the [`bias` ](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html#id36) preprocessor in the recipe instead and then plot the output of this simply with `1d_profile`. That will make your code here much simpler.

Thanks!! Let me know if you have any questions/comments!

valeriupredoi commented:
cheers both you guys! I can have a final look and merge when review's over and passed, remember to fix the [flake issues](https://app.circleci.com/pipelines/github/ESMValGroup/ESMValTool/9768/workflows/2c66e1f5-059c-484b-a12b-48d3fbb5f380/jobs/66742) and also please fix the PR description boxes (remove or strikethrough the ones that are not relevant, check mark the ones that are), yeah sorry, I sound like an accountant :grin:

FranziskaWinterstein commented:
So, finally I think I addressed all comments. @schlunma would you be so kind as to review the changes again? There are still some codacy issues - can these be ignored?

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -24,15 +24,26 @@
 
+      .. warning::
+
+      The plot_type ``profile`` for zonal mean profiles has been deprecated
+      in ESMValTool version 2.9.0 and is scheduled for removal in version
+      2.11.0. Please use plot type ``zonal_mean_profile`` instead. This is
+      an exact replacement.
```
```suggestion
          The plot_type ``profile`` for zonal mean profiles has been deprecated
          in ESMValTool version 2.9.0 and is scheduled for removal in version
          2.11.0. Please use plot type ``zonal_mean_profile`` instead. This is
          an exact replacement.
```

This should fix the documentation build

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -292,6 +304,43 @@
+
+Configuration options for plot type ``1d_profile``
+--------------------------------------------------
+legend_kwargs: dict, optional
+    Optional keyword arguments for :func:`matplotlib.pyplot.legend`. Use
+    ``legend_kwargs: false`` to not show legends.
+gridline_kwargs: dict, optional
+    Optional keyword arguments for grid lines. By default, ``color: lightgrey,
+    alpha: 0.5`` are used. Use ``gridline_kwargs: false`` to not show grid
+    lines.
+aspect_ratio: float, optional (default: 1.5)
+    Aspect ratio of the plot. The default value results in a slender upright
+    plot.
+log_x: bool, optional (default: False)
+    Use logarithmic X-axis.
+log_y: bool, optional (default: True)
+    Use logarithmic Y-axis.
+plot_kwargs: dict, optional
+    Optional keyword arguments for :func:`iris.plot.plot`. Dictionary keys are
+    elements identified by ``facet_used_for_labels`` or ``default``, e.g.,
+    ``CMIP6`` if ``facet_used_for_labels: project`` or ``historical`` if
+    ``facet_used_for_labels: exp``. Dictionary values are dictionaries used as
+    keyword arguments for :func:`iris.plot.plot`. String arguments can include
+    facets in curly brackets which will be derived from the corresponding
+    dataset, e.g., ``{project}``, ``{short_name}``, ``{exp}``. Examples:
+    ``default: {linestyle: '-', label: '{project}'}, CMIP6: {color: red,
+    linestyle: '--'}, OBS: {color: black}``.
+pyplot_kwargs: dict, optional
+    Optional calls to functions of :mod:`matplotlib.pyplot`. Dictionary keys
+    are functions of :mod:`matplotlib.pyplot`. Dictionary values are used as
+    single argument for these functions. String arguments can include facets in
+    curly brackets which will be derived from the datasets plotted in the
+    corresponding plot, e.g., ``{short_name}``, ``{exp}``. Facets like
+    ``{project}`` that vary between the different datasets will be transformed
+    to something like  ``ambiguous_project``. Examples: ``title: 'Awesome Plot
+    of {long_name}'``, ``xlabel: '{short_name}'``, ``xlim: [0, 5]``.
```
Could you please sort these options alphabetically? And I think `show_y_minor_ticklabels` is missing here, you can copy-paste it from the zonal-mean profile plot.

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -337,7 +386,7 @@ def __init__(self, config):
         super().__init__(config)
 
-        # Get default settings
+        # Get defaul stettings
```
```suggestion
        # Get default stettings
```

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -998,17 +1071,23 @@ def _process_pyplot_kwargs(self, plot_type, dataset):
                     f"pyplot_kwargs of {plot_type} '{func}: {arg}'",
                 )
-            getattr(plt, func)(arg)
+            if arg == 'None':
+                getattr(plt, func)()
+            else:
+                getattr(plt, func)(arg)
```
This could be problematic in some cases: `matplotlib` actually uses the string `None`. What is you use case here? If this is really necessary, it's probably better to check for the "true" `None` with `if arg is None:`. In the recipe, `None` can be specified with `null`.

FranziskaWinterstein replied:
Ah, okay, this was something I wanted to ask you. I had problems applying a pyplot_kwargs function with no argument. E.g. minorticks_on(). Do you have a suggest I was not aware of the `null` statement.

schlunma replied:
I see, yes, in that case you need to call it without arguments, so it's probably best to use

```py
if arg is None:
```

in the code and

```yml
pyplot_kwargs:
  minorticks_on: null
```

in the recipe.

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1360,14 +1440,115 @@ def create_profile_plot(self, datasets, short_name):
+        if self.plots[plot_type]['log_x']:
+            axes.set_xscale('log')
+            # major and minor ticks
+            x_major = LogLocator(base=10.0, numticks=12)
+            axes.get_xaxis().set_major_locator(x_major)
+            x_minor = LogLocator(base=10.0,
+                                 subs=np.arange(1.0, 10.0) * 0.1,
+                                 numticks=12)
+            axes.get_xaxis().set_minor_locator(x_minor)
+            axes.get_xaxis().set_minor_formatter(NullFormatter())
```
It's probably better to not hardcode the number of ticks here, you should be able to specify the tick locations in the recipe with

```yaml
pyplot_kwargs:
  xticks: [1, 2, 3]
```

(not tested though)

FranziskaWinterstein replied:
I actually had my issues here with the minor Log ticks. This seems an matplotlib issue (see [here](https://stackoverflow.com/questions/44078409/matplotlib-semi-log-plot-minor-tick-marks-are-gone-when-range-is-large) Would you be fine when I add numticks as an option with default 12. Those restrained minor ticks are only applied in the xlog case and can still be specified in the recipe for the non-xlog plots.

FranziskaWinterstein replied:
numticks is actually not changing the number of ticks, but need to be set equal or larger the number of displayed major ticks.

schlunma commented:
@esmvalbot please run monitor/recipe_monitor_with_refs.yml

valeriupredoi commented:
> Since @schlunma asked, ESMValBot will run recipe monitor/recipe_monitor_with_refs.yml as soon as possible, output will be generated [here](https://esmvaltool.dkrz.de/shared/esmvaltool/esmvalbot-output/monitor_1d_profile-lilt94yf)

yeah bot's on the ropes today, possibly due to Lustre FS issues at DKRZ, RIP poor bot :robot:

schlunma reviewed (approved):
I added the new plot type to our example recipe and update the doc, this should be ready now!

Thanks @FranziskaWinterstein again for this great addition!

@valeriupredoi would you do us the honor and perform a final check and merge please? I ran the recipe locally to produce the new plots for the doc, everything went well :rocket:

valeriupredoi commented:
great work here, guys! :beer: x2
